Branch-SDK/Branch-SDK/BNCServerInterface.m

```diff
@@ -133,70 +143,131 @@ - (void)genericHTTPRequest:(NSURLRequest *)request retryNumber:(NSInteger)retryN
+    [task resume];
+    [session finishTasksAndInvalidate];
 }
 
 - (BNCServerResponse *)genericHTTPRequest:(NSURLRequest *)request log:(BOOL)log {
     __block NSURLResponse *_response = nil;
     __block NSError *_error = nil;
     __block NSData *_respData = nil;
     
-    //NSURLSession is available in iOS 7 and above
-    if (NSFoundationVersionNumber >= NSFoundationVersionNumber_iOS_7_0) {
-        dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
-        NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
-        NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable urlResp, NSError * _Nullable error) {
-            _response = urlResp;
-            _error = error;
-            _respData = data;
-            dispatch_semaphore_signal(semaphore);
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    NSURLSession *session =
+        [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]
+            delegate:self
+            delegateQueue:nil];
+    NSURLSessionDataTask *task =
+        [session dataTaskWithRequest:request
+            completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable urlResp, NSError * _Nullable error) {
+                _response = urlResp;
+                _error = error;
+                _respData = data;
+                dispatch_semaphore_signal(semaphore);
         }];
-        [task resume];
-        [session finishTasksAndInvalidate];
-        dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
-    } else {
-        #pragma clang diagnostic push
-        #pragma clang diagnostic ignored "-Wdeprecated-declarations"
-        _respData = [NSURLConnection sendSynchronousRequest:request returningResponse:&_response error:&_error];
-        #pragma clang diagnostic pop
-    }
+    [task resume];
+    [session finishTasksAndInvalidate];
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
     return [self processServerResponse:_response data:_respData error:_error log:log];
 }
 
+- (void) URLSession:(NSURLSession *)session
+               task:(NSURLSessionTask *)task
+didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
+  completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition disposition, NSURLCredential *credential))completionHandler {
+
+    BOOL trusted = NO;
+    uint8_t kNonce[] = "Branch Rocks";
+    uint8_t encryptedString[2048];
+    size_t encryptedLength = sizeof(encryptedString);
+    SecTrustResultType trustResult = 0;
+    OSStatus err = 0;
+    NSData *data = nil;
+    NSString *string = nil;
+    NSString *hostName = nil;
+
+    // Release these:
+    SecKeyRef key = nil;
+    SecPolicyRef hostPolicy = nil;
+
+    NSString *encodedNonce =
```

Comment:
Not too sure the mechanics of all this, but could we also save a closely guarded "Branch Charles Proxy" cert that allowed our teams to debug?